### PR TITLE
Fix #1902 - Add --name option to create plugin

### LINKED ISSUE #1902: Add --name option to create plugin
##### Feature Request:

With `serverless create -t aws-nodejs --name my_service` you should be able to automatically overwrite the name of the created service

##### Benefits:

* Make it easier for users to set the name and not run into an issue when the name is set to the default name

# Linked Issues:
* #1616

### PR DESCRIPTION
**Pull request by mrlamroger**
##### Status:

**In Development**

##### Description:

Adds --name option to create plugin

Addresses https://github.com/serverless/serverless/issues/1902

##### Todos:

- [x] Write tests

### CODE REVIEW

**Comment by mrlamroger** [comment]
Right now, create copies the contents of the template directory through `serverless.utils.copyDirContentsSync`.

There are a couple of approaches I can see we try. 

1. Keep copying the contents. Read the serverless.yml file, replace the contents, and write it back

2. Don't just copy the contents. Template-tize the serverless.yml and replace the static template name with `options.name || options.template`. 

2 sounds like better bet imo. Less unnecessary copying. 1 would probably strip the comments out too.

**Comment by flomotlik** [comment]
@mrlamroger yup we removed the `--name` option basically because parsing the file removes the comments. Option 2 sounds good, as long as the comments stay in. I'd also be fine with a simple string replace (though having a nice templating system that automatically applies templates to potentially all files in the folder would be even cooler). 

Btw great job on writing the test first and opening a PR with it. Love this workflow!

**Comment by pmuens** [comment]
@mrlamroger thank you for your work so far! Also really in love with the TDD approach here! 🎉 

We've just introduced an update to the `create` plugin where we've added the `--path` functionality so that users can specify if a service should be created in a directory rather than the CWD (you can read the corresponding discussion here: https://github.com/serverless/serverless/issues/1726).

Furthermore we've decided to rename the service like the directory in which the service was created. So if I run `serverless create --template aws-nodejs --path my-service` a directory with the name `my-service` will be created (if not already present) and the service will be renamed to `my-service`.

The corresponding PR is this one: https://github.com/serverless/serverless/pull/1937

However we still have plans to allow the `--name` option for renaming the service. But we need to be careful here and discuss what should happen if someone provides a `--path` and `--name` option.

I would suggest that in this case the `--name` option wins.

Maybe we can wait until #1937 is merged and then go from there.

Any other thoughts @flomotlik @eahefnawy @mrlamroger

**Comment by flomotlik** [comment]
Yup name should overwrite the last part of the path imho or if you only do --name it will be created in the current directory and get the name set.

**Comment by mrlamroger** [comment]
Cool. Agreed @flomotlik and @pmuens. I'll add a test case for both cases after #1937 gets merged. 

I took a look at it too and noticed it does a string replace. I'll probably go the same route. A templating system seems out of scope for this ticket. We should revisit if `create.js` gets too bloated with case statement.

**Comment by flomotlik** [comment]
@mrlamroger #1937 was merged now, if you're interested in updating this that would be cool (if you don't have enough time at the moment to work on it let us know and we'll fix it up). Thanks for getting it to this point already!

**Comment by mrlamroger** [comment]
I'll fix tests in the AM. @flomotlik could you take a look please? #1937 did all the hard work. 
Also, the options dont reset between tests. =/

File: lib/plugins/create/create.js [hunk 1]
```diff
@@ -100,6 +105,11 @@ class Create {
       this.serverless.cli
         .log('NOTE: Please update the "service" property in serverless.yml with your service name');
     }
+    // if (!this.options.name) {
```
Comment (flomotlik) [inline lib/plugins/create/create.js]: Please remove the comment

File: lib/plugins/create/tests/create.js [hunk 2]
```diff
@@ -165,10 +181,13 @@ describe('Create', () => {
       process.chdir(tmpDir);
 
       create.options.path = 'my-new-service';
+      create.options.name = null;
 
       // using the nodejs template (this test is completely be independent from the template)
       create.options.template = 'aws-nodejs';
 
+      console.log("name is: " + create.options.name);
```
Comment (flomotlik) [inline lib/plugins/create/tests/create.js]: Remove please

File: lib/plugins/create/tests/create.js [hunk 3]
```diff
@@ -182,10 +201,44 @@ describe('Create', () => {
         const serverlessYmlfileContent = fse
           .readFileSync(path.join(serviceDir, 'serverless.yml')).toString();
 
+        console.log(serverlessYmlfileContent);
```
Comment (flomotlik) [inline lib/plugins/create/tests/create.js]: remove please

File: lib/plugins/create/tests/create.js [hunk 4]
```diff
@@ -182,10 +201,44 @@ describe('Create', () => {
...
+      fse.mkdirsSync(tmpDir);
+      process.chdir(tmpDir);
+
+      create.options.path = 'my-new-service';
+      create.options.name = 'my-custom-new-service';
+
+      // using the nodejs template (this test is completely be independent from the template)
+      create.options.template = 'aws-nodejs';
+
+      console.log("name is: " + create.options.name);
```
Comment (flomotlik) [inline lib/plugins/create/tests/create.js]: remove please

**Comment by flomotlik** [comment]
Thanks for the update. Few small changes but looks great in general.

File: lib/plugins/create/create.js [hunk 5]
```diff
@@ -100,6 +105,11 @@ class Create {
...
+    //   this.serverless.cli
```
Comment (pmuens) [inline lib/plugins/create/create.js]: Could you remove the commented out code as well?

File: lib/plugins/create/create.js (hunk 1 above)
Comment (mrlamroger) [inline lib/plugins/create/create.js]: Yep! That's embarrassing...

File: lib/plugins/create/create.js (hunk 5 above)
Comment (mrlamroger) [inline lib/plugins/create/create.js]: Yep!

**Comment by mrlamroger** [comment]
Updated with changes!

File: lib/plugins/create/create.js (hunk 5 above)
Comment (pmuens) [inline lib/plugins/create/create.js]: Thank you very much!

**Comment by mrlamroger** [comment]
What do you think @flomotlik ?

**Comment by flomotlik** [comment]
Great stuff, thanks for it!

**Pull request merged**